in_toto/verifylib.py

```
@@ -1043,8 +1044,105 @@ def reduce_chain_links(chain_link_dict):
 
   return reduced_chain_link_dict
 
+def verify_delegations(layout, chain_link_dict):
+  """
+  <Purpose>
+    Checks if any step has been delegated by the functionary, recurses into
+    the delegation and replaces the layout object in the chain_link_dict
+    by an equivalent link object
+
+  <Arguments>
+    layout:
+            The layout specified by the project owner against which the
+            threshold will be verified.
+
+    chain_link_dict:
+            A dictionary of key-link pair with step names as keys. For each
+            step name, there are one or more keyids and corresponding
+            link objects.
+
+  <Exceptions>
+    raises an Exception if verification of the delegated step fails.
+
+  <Side Effects>
+    None.
 
-def in_toto_verify(layout_path, layout_key_paths):
+  """
+
+  for step_name, key_link_dict in six.iteritems(chain_link_dict):
+
+    for keyid, link in six.iteritems(key_link_dict):
+
+      # Extract keyid and link from the passed chain_link_dict
+      # Check if the link object is of type 'layout'
+      if link._type == "layout":
+        layout_key_dict = {}
+
+        for key, key_obj in six.iteritems(layout.keys):
+
+          # Extract the key object from the passed layout
+          if key == keyid:
+            # Extract the key object of the key corresponding
+            # to the link object
+            layout_key_dict[key] = key_obj
```

I think you can replace lines 1079:1087 with one line of code:
```python
layout_key_dict = {keyid: layout.keys.get(keyid)}
```
and a good comment about what it means using that key.

---

I imagine it'd be also necessary to check if layout_key_dict is not none in case this change happens...

---

We could, but we don't have to, the format will be checked down the road.

---

Right, and it would fail with a "this layout is of the wrong format" or something of that nature. I'm usually for failing early.

---

@SantiagoTorres I’m not sure if we really need it. The layout_key_dict would be None if we are not able to fetch the key object from the layout. But if the key object is not present in the layout we wouldn’t reach this step. Because we verify the layout signatures and step signatures before this step where we retrieve the key object from the layout. So if the key object is not present, we would fail early.